### PR TITLE
[Order] Order any param (Nullable or with default value) that has type of Object

### PR DESCRIPTION
**Pull request by dobryy**

### CODE REVIEW

File: rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php [hunk 1]
```diff
@@ -83,24 +83,26 @@ private function getSortedParams(ClassMethod $classMethod): array
...
-            $typeA = $firstParam->type;
-            /** @var Name $typeB */
-            $typeB = $secondParam->type;
-            return strcmp($this->getShortName($typeA), $this->getShortName($typeB));
+            /** @var Name $firstParamType */
+            $firstParamType = $this->getParamType($firstParam);
+            /** @var Name $secondParamType */
+            $secondParamType = $this->getParamType($secondParam);
+
+            return strcmp($this->getShortName($firstParamType), $this->getShortName($secondParamType));
```
Comment (dobryy) [inline rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php:91]: @TomasVotruba Just curious... is it okay that I'm making phpstan fool here? :D
`getParamType` returns `Identifier|Name|UnionType|null` but in this method I'm changing the type in annotation to `Name`

Comment (TomasVotruba) [reply rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php:91]: I would not say fool, but rather share with him what you know :)
If that's 100 % the case there is a Name, then why not :+1: 

In time there might be an exception, but that's what mostly running rule on Rector code base revels.

Comment (TomasVotruba) [reply rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php:91]: What about adding test case completely without type?

```php
public function __construct($whatever, stdClass $someObject)
{
}
```

Comment (dobryy) [reply rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php:91]: It is already there ;)

Comment (dobryy) [reply rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php:91]: https://github.com/rectorphp/rector/blob/ec6a9d8da05caecf2e3da1752c67ea02dc89d9ca/rules/order/tests/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector/Fixture/skip_param_with_no_type.php.inc

Comment (TomasVotruba) [reply rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php:91]: Btw, how does `strcmp` compare to `<=>`? I never used `strcmp`, as the name is not clear what it does for me, so I'm curious

Comment (dobryy) [reply rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php]: Changed to `<=>`

**Comment by TomasVotruba** [comment]
Looks good :+1: 
Just rebase on current `master` and it's ready to merge.

When do you plan to apply rule on Rector and add it to CI? I wonder how many places it discovers in Rector.

**Comment by dobryy** [comment]
> Looks good 
> Just rebase on current `master` and it's ready to merge.
Ready

> When do you plan to apply rule on Rector and add it to CI? I wonder how many places it discovers in Rector.

Actually, I applied it locally since it wasn't set up in CI (I shouldn't?):
https://github.com/rectorphp/rector/pull/3795/commits/1f7b2d88d09e7acfb43bf54bf32bfb5f601e60f1
https://github.com/rectorphp/rector/pull/3786/commits/f37cd4cac367c3522497cf1c55fb3ee2f2d5a271

**Comment by dobryy** [comment]
@TomasVotruba, I'm going to enable all Order rules except OrderPropertyByComplexityRector (this one throws errors) in this PR https://github.com/rectorphp/rector/pull/3791. Then I'll take a look at the OrderPropertyByComplexityRector rule.

**Comment by TomasVotruba** [comment]
Looks good, let's :ship: it!

**Pull request merged**